### PR TITLE
Fix json returned from query REST entripoint

### PR DESCRIPTION
**Pull request by fax4ever**
There are stuff we can do very much better...

### CODE REVIEW

File: manager/src/test/java/io/gingersnapproject/rules/JoinTest.java [hunk 1]
```diff
@@ -56,9 +58,12 @@ public void testJoin() throws Exception {
 
       Thread.sleep(1000); // we need to wait for the ES near real time behavior
 
+      StringExtractorMatcher stringExtractor = new StringExtractorMatcher();
       given().queryParam("query", "select * from flight where name = 'BA0666' order by scheduled_time")
             .when().get(RULE_PATH)
-            .then().body(containsString("hitCount=1"));
+            .then().body(stringExtractor);
```
Comment (ryanemerson) [inline manager/src/test/java/io/gingersnapproject/rules/JoinTest.java]: We should be able to just reference the json fields directly via body, e.g:

```java
.body("hitCount", equalTo(1))
.body("hits[0].name", equalTo("BA0666"));
```

Comment (ryanemerson) [reply manager/src/test/java/io/gingersnapproject/rules/JoinTest.java]: I've added a commit to your fork :+1:

File: manager/src/main/java/io/gingersnapproject/search/QueryHandler.java [hunk 2]
```diff
@@ -24,9 +24,7 @@ public Uni<QueryResult> query(String query) {
             // Should we allow concurrency?
             .concatenate();
 
-      return result
-            .onItem()
-            .transform(searchResult -> new QueryResult(searchResult.hitCount(), searchResult.hitCountExact(),
-                  hits, searchResult.hitsExact()));
+      return Uni.combine().all().unis(result, hits.collect().asList()).combinedWith((searchResult, strings) ->
+            new QueryResult(searchResult.hitCount(), searchResult.hitCountExact(), strings, searchResult.hitsExact()));
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/search/QueryHandler.java]: This code should probably be
```java
   public Uni<QueryResult> query(String query) {
      Uni<SearchResult> result = searchBackend.get().query(query);
      return result.onItem().transformToUni(sr ->
         Multi.createFrom().iterable(sr.hits())
               .onItem().transformToUni(sh -> caches.get(sh.indexName(), sh.documentId()))
               .concatenate()
               .collect().asList()
               .onItem().transform(list -> new QueryResult(sr.hitCount(), sr.hitCountExact(), list, sr.hitsExact()))
      );
   }
```
as it describes the relationship a bit better. The combine operator is more for separate Unis, where as the hits are not available until the result Uni has completed which is more clear with this code.

File: manager/src/main/java/io/gingersnapproject/json/JsonHelper.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+         result.set("hitsExacts", queryResult.hitsExacts());
+
+         return result.toString();
+      });
+   }
+
+   public static Json make(List<String> hits) {
+      Json array = Json.array();
+      for (String item : hits) {
+         // TODO this is very inefficient
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/json/JsonHelper.java:33]: I don't think there is any other way to do this if you want a nested Json object. It needs to be parsed. Or are you thinking of creating a larger String array Json representation and calling to parse once?

File: manager/src/main/java/io/gingersnapproject/json/JsonHelper.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import io.smallrye.mutiny.Uni;
+
+public final class JsonHelper {
+
+   private JsonHelper() {
+   }
+
+   public static Uni<String> json(Uni<QueryResult> queryResultUni) {
+      return queryResultUni.onItem().transform(queryResult -> {
+         // Json.make() seems to be not working:
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/json/JsonHelper.java:17]: What isn't working exactly?

File: manager/src/main/java/io/gingersnapproject/json/JsonHelper.java (hunk 3 above)
Comment (fax4ever) [reply manager/src/main/java/io/gingersnapproject/json/JsonHelper.java:33]: Maybe we could avoid parsing the JSON hits, building the entire QueryResult JSON manually from a StringBuilder. But in any case let's keep it as it is for now.

File: manager/src/main/java/io/gingersnapproject/search/QueryHandler.java (hunk 2 above)
Comment (fax4ever) [reply manager/src/main/java/io/gingersnapproject/search/QueryHandler.java]: Much better thanks. I'm applying it.

File: manager/src/main/java/io/gingersnapproject/json/JsonHelper.java (hunk 4 above)
Comment (fax4ever) [reply manager/src/main/java/io/gingersnapproject/json/JsonHelper.java:17]: If I try:
``` java
// Json.make() seems to be not working:
// > doing it manually
Json result = Json.make(queryResult);
```

I get the error:
``` bash
Don't know how to convert to Json : QueryResult[hitCount=1, hitCountExact=true, hits=[{"name":"BA0666","scheduled_time":"12:00:00","airline_id":{"iata":"BA","name":"British Airways"},"gate_id":{"name":"B1"}}], hitsExacts=true]
```

**Comment by fax4ever** [comment]
Thank @wburns I applied the changes.

**Comment by ryanemerson** [comment]
Thanks @fax4ever

**Pull request merged**